src/WattleScript.Interpreter/Tree/Statements/ForEachLoopStatement.cs

```diff
@@ -164,6 +172,71 @@ public override void Compile(FunctionBuilder bc)
 			bc.PopSourceRef();
 		}
 
+		private bool CompileRangeStatement(Loop l, FunctionBuilder bc)
+		{
+			if (!(m_RValues is ExprListExpression listExpr)) return false;
+			if (listExpr.Expressions.Count != 1) return false;
+			
+			Expression expr = listExpr.Expressions[0];
+
+			if (!(expr is BinaryOperatorExpression binaryExpr)) return false;
+			if (!binaryExpr.IsRangeCtor()) return false;
+			
+			if (binaryExpr.Operator == Operator.RightExclusiveRange || binaryExpr.Operator == Operator.ExclusiveRange) // ..<, >..< -> sub 1
+			{
+				if (binaryExpr.Exp2 is LiteralExpression le2 && le2.Value.Type == DataType.Number)
```

Comment:
The correct way to do things with literal values is to use the `EvalLiteral` function to get the final value. This lets expressions like `(1 + 2)` or `2 ** 4` be passed through as literals. It also makes sense to use TryCastToNumber here to allow for that string coercion.

Finally, when it _isn't_ a literal we do still need those -1 and +1 opcodes to be emitted. Add a test that references variables, arguments to make sure this still holds.

Example:

```cs
if(binaryExpr.Exp2.EvalLiteral(out var le2) && le2.TryCastToNumber(out var leftNum)) {
    bc.Emit_Literal(DynValue.NewNumber((int)(leftNum - 1)));
} else {
    binaryExpr.Exp2.Compile(bc);
    bc.Emit_Literal(DynValue.NewNumber(1));
    bc.Emit_Operator(OpCode.Sub);
}
```

Reply:
_Should be_ fixed in 8895379822ee7dbcd73e75bfd4dbb369a4f0e929, please take a look whether something else needs to be done. Added a bunch of tests